keps/sig-api-machinery/2885-server-side-unknown-field-validation/README.md

```diff
@@ -0,0 +1,913 @@
+opt-in to server side schema validation).
+
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+### Opt-in API Mechanism
+
+There are a few ways we could allow a client to opt-in to server-side unknown
+field validation, we present two options **Content-Type Header** and **query
+parameter**.
+
+#### Content-Type Header
+
+Requests that contain a kubernetes object, also pass along with it a
+content-type header such as “application/json”. One way to indicate to the
+server that it should use strict schema validation and fail when unknown fields
+are passed is to send a mime-type parameter indicating the strict validation
+such as “application/json;validation=strict”. Alternatively we could use a new
+header such as “X-Kubernetes-Validation:strict”.
+
+One could argue that this is the more appropriate way to parameterize opting-in
+to server side schema validation because on the server we are fundamentally
+treating the content as a different type (“application/json” is json data that
+we don’t care if it has extra fields, “application/json;validation=strict” is
+json data that is sensitive to extra fields).
+
+A precedent for using a mime-type parameter is from how we [receive resources as
+tables](https://kubernetes.io/docs/reference/using-api/_print/#receiving-resources-as-tables).
+
+On the other hand, one could also argue that interpreting input strictly or not
+according to the target schema is independent of the content type and that this
+is an inappropriate way to parameterize validation opt-in.
+
+Another argument against this method is that for patch, strictness is handled
+when decoding the *result* of the patch, so it wouldn’t make sense to use
+Content-Type which is providing information about the inbound request body.
+
+#### Query Parameter
+
+Alternatively, if we don’t like the idea of using Content-Type header to
+determine whether the apiserver should accept or fail on unknown fields, we
+could pass a query param such as “?validate=true”.
+
+This might make it more obvious to consumers of the API that strict schema
+validation is a choice of the client. On the other hand, query parameters are
+more typically used for filtering/sorting data returned from the API server.
+
+Arguments for using query parameter include:
+* Being able version them with the API version (unlike content-type which must
+  be versioned with the type)
+* Parameters are more discoverable in openapi, less so for mime types.
+* We have precedent for using query parameters for write requests already (via CreateOptions,
+  PatchOptions, UpdateOptions)
+
+
+We believe that using a query parameter is the best approach.
```

Comment:
What will the client-go support look like? Will the param be added to CreateOptions and the like?

Reply:
Yep, that's correct. I've added a note to the query param opt-in section.